subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+	}
+
+	req = net_buf_simple_pull_mem(buf, sizeof(*req));
+
+	preset = preset_get(req->idx);
+	if (!preset) {
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+
+	if (!(preset->properties & BT_HAS_PROP_AVAILABLE)) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	return call_ops_set_active_preset(preset->id, sync);
+}
+
+static bool find_available(struct has_preset *preset, void *user_data)
+{
+	struct has_preset **found = user_data;
+
+	if ((preset->properties & BT_HAS_PROP_AVAILABLE) && !preset->hidden) {
+		*found = preset;
+		return true;
+	}
+
+	return false;
+}
+
+static int handle_set_next_preset(struct has_client *client, struct net_buf_simple *buf, bool sync)
+{
+	struct has_preset *preset = NULL;
+
+	preset_foreach(has_local.active_id + 1, last_preset_id, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	preset_foreach(0x01, has_local.active_id - 1, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+}
+
+static int handle_set_prev_preset(struct has_client *client, struct net_buf_simple *buf, bool sync)
+{
+	struct has_preset *preset = NULL;
+
+	preset_foreach(0x01, has_local.active_id - 1, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	preset_foreach(has_local.active_id + 1, last_preset_id, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+}
+
+static ssize_t control_point_rx(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *data, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	const struct bt_has_cp_hdr *hdr;
+	struct has_client *client;
+	struct net_buf_simple buf;
+	ssize_t ret;
+
+	client = client_find(conn);
+	if (!client) {
+		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+	}
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len < sizeof(*hdr)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)data, len);
+
+	hdr = net_buf_simple_pull_mem(&buf, sizeof(*hdr));
+
+	BT_DBG("conn %p op %s (0x%02x)", conn, bt_has_op_str(hdr->op), hdr->op);
+
+	switch (hdr->op) {
+	case BT_HAS_OP_READ_PRESET_REQ:
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
+			return BT_GATT_ERR(BT_ATT_ERR_CCC_IMPROPER_CONF);
+		}
+
+		/* Fail if ATT MTU is potentially insufficient to complete the operation */
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_ATT_MTU_VALID)) {
+			return BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+		}
+
+		ret = handle_read_preset_req(client, &buf);
+		break;
+
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+	case BT_HAS_OP_WRITE_PRESET_NAME:
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
```

Comment:
Because indications are mandatory for the HARC.
But is is mandatory for the server to ensure that the client is well-behaving?
Probably better to give error message, as is done here, than to accept and not get response through.

Reply:
> Because indications are mandatory for the HARC. But is is mandatory for the server to ensure that the client is well-behaving? Probably better to give error message, as is done here, than to accept and not get response through.

HAS spec says:
"If a client writes 
    • The Read Presets Request opcode, or
    • The Write Preset Name opcode, or
    • The Set Active Preset opcode, or
    • The Set Active Preset – Synchronized Locally opcode
to the Hearing Aid Control Point characteristic, and the Client Characteristic Configuration Descriptor of the Hearing Aid Control Point is not configured for indications, the server shall return an ATT_ERROR_RSP PDU to the ATT_WRITE_REQ PDU with the error code parameter set to Client Characteristic Configuration Descriptor Improperly Configured as defined in [3]."